### PR TITLE
llext: add tests for init arrays

### PR DESCRIPTION
**Pull request by pillo79**
This PR is based on #77641 (first 3 commits) and adds the test suite for that feature. Will be set as ready as soon as the parent PR is merged.

LLEXT test case definition has been improved to use struct syntax, so that tests can easily define many custom features. New `test_cleanup` callback has been added to check the proper execution of the routine linked in `.fini_array`.

### CODE REVIEW

File: tests/subsys/llext/simple/src/init_fini_ext.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	init_fn
+};
+static const void *const fini_fn_ptrs[] __used Z_GENERIC_SECTION(".fini_array") = {
+	fini_fn
+};
+
+void test_entry(void)
+{
+	/* fini_fn() is not called yet, so we expect 0x123 */
+	const uint32_t expected = (((1 << 4) | 2) << 4) | 3;
```
Comment (lyakh) [inline tests/subsys/llext/simple/src/init_fini_ext.c:66]: why not just `int`?

File: tests/subsys/llext/simple/src/test_llext_simple.c [hunk 2]
```diff
@@ -176,7 +167,7 @@ void load_call_unload(const struct llext_test *test_case)
 	k_thread_join(&llext_thread, K_FOREVER);
 
 	if (test_case->test_cleanup) {
-		test_case->test_cleanup(&ext);
+		test_case->test_cleanup(ext);
```
Comment (lyakh) [inline tests/subsys/llext/simple/src/test_llext_simple.c:170]: merge into previous one

File: tests/subsys/llext/simple/src/test_llext_simple.c [hunk 3]
```diff
@@ -251,6 +255,29 @@ LLEXT_LOAD_UNLOAD(hello_world,
...
+	#include "init_fini.inc"
+};
+
+static void init_fini_test_cleanup(struct llext *ext)
+{
+	/* Make sure fini_fn() was called during teardown.
+	 * (see init_fini_ext.c for more details).
+	 */
+	const uint32_t *number = llext_find_sym(&ext->exp_tab, "number");
+	const uint32_t expected = (((((1 << 4) | 2) << 4) | 3) << 4) | 4; /* 0x1234 */
```
Comment (lyakh) [inline tests/subsys/llext/simple/src/test_llext_simple.c:269]: `int`?

File: tests/subsys/llext/simple/src/test_llext_simple.c (hunk 2 above)
Comment (pillo79) [reply tests/subsys/llext/simple/src/test_llext_simple.c:170]: Botched the fixups! Thanks for spotting this :+1:

File: tests/subsys/llext/simple/src/init_fini_ext.c (hunk 1 above)
Comment (pillo79) [reply tests/subsys/llext/simple/src/init_fini_ext.c:66]: No special reason. We also need to be 64-bit inclusive now :smile: so I'll switch.

**Review by teburd — DISMISSED** [review]

**Review by teburd — APPROVED** [review]

**Review by mathieuchopstm — APPROVED** [review]
LGTM. Few nits if you need to make changes:

**Pull request merged**